implements hellinger distance as a drift calculation

FelipeAdachi opened the pull request:
## Description

Implements hellinger distance to be applied between two profile views.

Currently, this is meant to be used as a standalone function to get the Hellinger Distance statistic. Further work must be done to integrate it with the `viz` module, and the decision process of which kind of algorithm/drift test is applied in different circumstances.

This currently enables hellinger for any feature that has a Distribution Metric. FI Metrics-only features are still not supported.

This repository needs <!-- Add 1-2 sentences explaining the purpose of this PR. -->

The commits in this pull request will <!-- Summarize PR. -->

## Changes

- Imperative commit or change title (commit ID) <!-- Add commit ID and [GitHub will autolink](https://help.github.com/en/github/writing-on-github/autolinked-references-and-urls). -->
  - Add some bullet points to explain the change
- Next commit or change (commit ID)

## Related

Relates to organization/repo#number

<!-- Reference related commits, issues and pull requests. Type `#` and select from the list. -->

Closes organization/repo#number

<!-- List issues for GitHub to automatically close after merge to default branch. Type `#` and select from the list. See the [GitHub docs on linking PRs to issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) for more. -->

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

FelipeAdachi committed:
implements hellinger distance as a drift calculation

richard-rogers reviewed (dismissed):
LGTM

jamie256 commented on python/whylogs/viz/utils/drift_calculations.py:
```diff
@@ -1,16 +1,24 @@
+from whylogs.viz.utils.histogram_calculations import (
+    HIST_AVG_NUMBER_PER_BUCKET,
+    MAX_HIST_BUCKETS,
+    _calculate_bins,
```
Consider moving those constants and `_calculate_bins` to the drift_calculations.py file (or some common location). One concern is that we may require other extras in order to use this method programmatically?

FelipeAdachi replied:
done in the latest commits. we can open another one if something needs to be done around this comment:

`One concern is that we may require other extras in order to use this method programmatically?`

jamie256 commented on python/whylogs/viz/utils/drift_calculations.py:
```diff
@@ -183,9 +198,40 @@ def _get_chi2_p_value(target_view_column, reference_view_column) -> Optional[Col
+    target_pmf = target_kll_sketch.get_pmf(bins)
+    ref_pmf = ref_kll_sketch.get_pmf(bins)
+
+    distance = euclidean(np.sqrt(target_pmf), np.sqrt(ref_pmf)) / np.sqrt(2)
```
Might be worth factoring the hellinger distance calculation out into a separate function for better reusability (signature could be two pmfs rather than ColumnProfileViews), can you also add a comment with reference link to the equation used? e.g. https://en.wikipedia.org/wiki/Hellinger_distance

I like that you explicitly use euclidean here as it makes this easier to understand.

Do you need to add numpy as a dependency of whylogs[viz] with this change?

FelipeAdachi replied:
Yes, absolutely, will do that!

oh wow, I haven't realized that numpy is not already a core dependency. If that's the case, then yes, it should be added as a dependency to whylogs[viz]

FelipeAdachi replied:
done in the latest commit

jamie256 reviewed (commented):
Very nice to see this get into our drift detection algorithms! Thanks!

FelipeAdachi committed:
- centralize constants at a config file for viz module
- separate hellinger distance calculation to make it more generalizable
- better annotation and documentation

FelipeAdachi committed:
isort

jamie256 reviewed (dismissed):
LGTM!

bernease commented on python/whylogs/viz/utils/drift_calculations.py:
```diff
@@ -183,9 +196,62 @@ def _get_chi2_p_value(target_view_column, reference_view_column) -> Optional[Col
+        Between 0 (identical distributions) and 1 (completely different distributions).
+    """
+    # https://en.wikipedia.org/wiki/Hellinger_distance
+    distance = euclidean(np.sqrt(target_pmf), np.sqrt(reference_pmf)) / np.sqrt(2)
```
This should be divided by a constant sqrt(2).

bernease commented on python/whylogs/viz/utils/drift_calculations.py:
```diff
@@ -183,9 +196,62 @@ def _get_chi2_p_value(target_view_column, reference_view_column) -> Optional[Col
+    start = min([target_kll_sketch.get_min_value(), ref_kll_sketch.get_min_value()])
+    end = max([target_kll_sketch.get_max_value(), ref_kll_sketch.get_max_value()])
+    n = target_kll_sketch.get_n() + ref_kll_sketch.get_n()
+    bins, end = _calculate_bins(end=end, start=start, n=n, avg_per_bucket=HIST_AVG_NUMBER_PER_BUCKET, max_buckets=nbins)
```
This choosing the number of buckets based on an average number of values per bucket (up to some maximum number of buckets) is new to me. Could we experiment with this using the tests that you're running? I don't see anything wrong with it offhand, but it's just unfamiliar.

FelipeAdachi replied:
Yes, that's unfamiliar to me as well, this was carried from v0.

Actually, none of the experiments I ran fall into a case where this avg_per_bucket would be used. But I can run tests to specifically hit that scenario.

FelipeAdachi replied:
I will make tests and experiments on this on a follow-up

bernease reviewed (changes requested):
Mostly minor changes, but please do multiply the Hellinger distance by 1/sqrt(2)!

FelipeAdachi committed:
Update drift_calculations.py

annotating nbins

bernease reviewed (approved):
Updating PR to approved because I can not read.
